segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/cache/DirectMemoryCache.java

```diff
@@ -0,0 +1,437 @@
+
+            for (DirectMemoryBuffer b : this.buffers) {
+                b.close();
+            }
+        }
+    }
+
+    //endregion
+
+    //region CacheStorage Implementation
+
+    @Override
+    public int getBlockAlignment() {
+        return this.layout.blockSize();
+    }
+
+    @Override
+    public int getMaxEntryLength() {
+        return CacheLayout.MAX_ENTRY_SIZE;
+    }
+
+    @Override
+    public int insert(BufferView data) {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        Preconditions.checkArgument(data.getLength() <= CacheLayout.MAX_ENTRY_SIZE,
+                "Entry too long. Expected max %s, given %s.", CacheLayout.MAX_ENTRY_SIZE, data.getLength());
+
+        int lastBlockAddress = CacheLayout.NO_ADDRESS;
+        int remainingLength = data.getLength();
+        try {
+            // As long as we still have data to copy, we try to reuse an existing, non-full buffer or allocate a new one,
+            // and write the remaining data to it.
+            while (remainingLength > 0 || lastBlockAddress == CacheLayout.NO_ADDRESS) {
+                // Get a Buffer to write data to. If we are full, this will throw an appropriate exception.
+                DirectMemoryBuffer buffer = getNextAvailableBuffer();
+
+                // Write the data to the buffer.
+                BufferView slice = data.slice(data.getLength() - remainingLength, remainingLength);
+                DirectMemoryBuffer.WriteResult writeResult = buffer.write(slice, lastBlockAddress);
+                if (writeResult == null) {
+                    // Someone else grabbed this buffer and wrote to it before we got a chance. Go back and find another one.
+                    continue;
+                }
+
+                // Update our state, including the number of bytes written. In case of a subsequent error (and rollback),
+                // invoking delete() will undo this changes as well.
+                assert writeResult.getWrittenLength() >= 0 && writeResult.getWrittenLength() <= remainingLength : writeResult.getWrittenLength();
+                remainingLength -= writeResult.getWrittenLength();
+                this.storedBytes.addAndGet(writeResult.getWrittenLength());
+                lastBlockAddress = writeResult.getLastBlockAddress();
+            }
+        } catch (Throwable ex) {
+            if (!Exceptions.mustRethrow(ex) && lastBlockAddress != CacheLayout.NO_ADDRESS) {
+                // We wrote something, but got interrupted. We need to clean up whatever we wrote so we don't leave
+                // unreferenced data in the cache.
+                delete(lastBlockAddress);
+            }
+            throw ex;
+        }
+
+        CacheMetrics.insert(data.getLength());
+        return lastBlockAddress;
+    }
+
+    @Override
+    public int replace(int address, BufferView data) {
+        // The simplest and safest way to do a replace is to first insert the new data and then remove the old one. This
```

Comment:
Given this implementation, is this API really providing value? Couldn't a caller just do that.

Reply:
Yes, given the **current** implementation, the caller can just do the same. However this API has a drawback in that it requires extra (temporary) cache storage space. Both the old and the new entry must be stored in the cache prior to deleting the old one.

In the future, we may want to optimize this to replace the entry block-by block, in which case it won't require these 2 steps. That may also provide a small perf gain.

I want to keep this so that we may implement any optimizations we want inside the cache and keep track of those callsites that use replacement.